clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsRebalanceEventsProcessor.java

```diff
@@ -0,0 +1,205 @@
+     * Sets the application event handler.
+     *
+     * The application handler sends the results of the callbacks to the background thread.
+     *
+     * @param applicationEventHandler The application handler.
+     */
+    public void setApplicationEventHandler(final ApplicationEventHandler applicationEventHandler) {
+        this.applicationEventHandler = applicationEventHandler;
+    }
+
+    private void process(final BackgroundEvent event) {
+        switch (event.type()) {
+            case ERROR:
+                throw ((ErrorEvent) event).error();
+
+            case STREAMS_ON_TASKS_REVOKED_CALLBACK_NEEDED:
+                processStreamsOnTasksRevokedCallbackNeededEvent((StreamsOnTasksRevokedCallbackNeededEvent) event);
+                break;
+
+            case STREAMS_ON_TASKS_ASSIGNED_CALLBACK_NEEDED:
+                processStreamsOnTasksAssignedCallbackNeededEvent((StreamsOnTasksAssignedCallbackNeededEvent) event);
+                break;
+
+            case STREAMS_ON_ALL_TASKS_LOST_CALLBACK_NEEDED:
+                processStreamsOnAllTasksLostCallbackNeededEvent((StreamsOnAllTasksLostCallbackNeededEvent) event);
+                break;
+
+            default:
+                throw new IllegalArgumentException("Background event type " + event.type() + " was not expected");
+
+        }
+    }
+
+    private void processStreamsOnTasksRevokedCallbackNeededEvent(final StreamsOnTasksRevokedCallbackNeededEvent event) {
+        StreamsOnTasksRevokedCallbackCompletedEvent invokedEvent = invokeOnTasksRevokedCallback(event.activeTasksToRevoke(), event.future());
+        applicationEventHandler.add(invokedEvent);
+        if (invokedEvent.error().isPresent()) {
+            throw invokedEvent.error().get();
+        }
+    }
+
+    private void processStreamsOnTasksAssignedCallbackNeededEvent(final StreamsOnTasksAssignedCallbackNeededEvent event) {
+        StreamsOnTasksAssignedCallbackCompletedEvent invokedEvent = invokeOnTasksAssignedCallback(event.assignment(), event.future());
+        applicationEventHandler.add(invokedEvent);
+        if (invokedEvent.error().isPresent()) {
+            throw invokedEvent.error().get();
+        }
+    }
+
+    private void processStreamsOnAllTasksLostCallbackNeededEvent(final StreamsOnAllTasksLostCallbackNeededEvent event) {
+        StreamsOnAllTasksLostCallbackCompletedEvent invokedEvent = invokeOnAllTasksLostCallback(event.future());
+        applicationEventHandler.add(invokedEvent);
+        if (invokedEvent.error().isPresent()) {
+            throw invokedEvent.error().get();
+        }
+    }
+
+    private StreamsOnTasksRevokedCallbackCompletedEvent invokeOnTasksRevokedCallback(final Set<StreamsRebalanceData.TaskId> activeTasksToRevoke,
+                                                                                     final CompletableFuture<Void> future) {
+        final Optional<Exception> exceptionFromCallback = rebalanceCallbacks.onTasksRevoked(activeTasksToRevoke);
```

Comment:
`invokeOnTasksRevokedCallback` has a different implementation than `invokeOnTasksAssignedCallback` and `invokeOnAllTasksLostCallback`. It's probably easier to read if all three use either the `.map` stile or the `isPresent` style. I don't mind which one, `isPresent` is probably a little easier to follow.

Reply:
In this case I just applied IDEA's recommendation to transform it to functional style. However, I agree with you regarding readability.